crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -212,21 +212,29 @@ impl ClassInfoConstraintFunction {
                 )
             }),
 
-            Type::KnownInstance(KnownInstanceType::UnionType(elements)) => {
-                UnionType::try_from_elements(
-                    db,
-                    elements.elements(db).iter().map(|element| {
-                        // A special case is made for `None` at runtime
-                        // (it's implicitly converted to `NoneType` in `int | None`)
-                        // which means that `isinstance(x, int | None)` works even though
-                        // `None` is not a class literal.
-                        if element.is_none(db) {
-                            self.generate_constraint(db, KnownClass::NoneType.to_class_literal(db))
-                        } else {
-                            self.generate_constraint(db, *element)
-                        }
-                    }),
-                )
+            Type::KnownInstance(KnownInstanceType::UnionType(instance)) => {
+                match instance {
+                    UnionTypeInstance::Legacy { .. } => None,
```

Comment:
Will look into this tomorrow, but I guess a legacy `Union[int, str]` is not allowed in `isinstance`/`issubclass`. If so, we were definitely not handling it correctly before.

Reply:
It actually _is_ allowed in isinstance/issubclass, but only on Python 3.10+. And I wouldn't worry too much about older versions than that since they're now end-of-life.

We should probably add some tests for this; it's an edge case but it has come up a couple of times on the mypy issue tracker.